src/libsync/pushnotifications.cpp

```diff
@@ -165,33 +175,128 @@ void PushNotifications::handleAuthenticated()
     qCInfo(lcPushNotifications) << "Authenticated successful on websocket";
     _failedAuthenticationAttemptsCount = 0;
     _isReady = true;
+    startPingTimeoutTimer();
     emit ready();
+
+    // We maybe reconnected to websocket while being offline for a
+    // while. To not miss any notifications that may have happend,
+    // emit all the signals once.
+    emitFilesChanged();
+    emitNotificationsChanged();
+    emitActivitiesChanged();
 }
 
 void PushNotifications::handleNotifyFile()
 {
     qCInfo(lcPushNotifications) << "Files push notification arrived";
-    emit filesChanged(_account);
+    emitFilesChanged();
 }
 
 void PushNotifications::handleInvalidCredentials()
 {
     qCInfo(lcPushNotifications) << "Invalid credentials submitted to websocket";
     if (!tryReconnectToWebSocket()) {
-        _isReady = false;
+        closeWebSocket();
         emit authenticationFailed();
     }
 }
 
 void PushNotifications::handleNotifyNotification()
 {
     qCInfo(lcPushNotifications) << "Push notification arrived";
-    emit notificationsChanged(_account);
+    emitNotificationsChanged();
 }
 
 void PushNotifications::handleNotifyActivity()
 {
     qCInfo(lcPushNotifications) << "Push activity arrived";
+    emitActivitiesChanged();
+}
+
+void PushNotifications::onWebSocketPongReceived(quint64 /*elapsedTime*/, const QByteArray &payload)
+{
+    handleTimeoutPong(payload);
+}
+
+void PushNotifications::handleTimeoutPong(const QByteArray &payload)
+{
+    // We are not interested in different pongs from the server
+    const auto expectedPingPayload = timeoutPingPayload();
+    if (payload != expectedPingPayload) {
+        return;
+    }
+
+    qCDebug(lcPushNotifications) << "Pong received in time";
+
+
+    _timeoutPongReceivedFromWebSocketServer = true;
+    startPingTimeoutTimer();
+}
+
+void PushNotifications::startPingTimeoutTimer()
+{
+    _pingTimeoutTimer.stop();
+    _pingTimer.setInterval(_pingTimeoutInterval);
+    _pingTimer.setSingleShot(true);
+    connect(&_pingTimer, &QTimer::timeout, this, &PushNotifications::pingWebSocketServer, Qt::UniqueConnection);
+    _pingTimer.start();
+}
+
+void PushNotifications::startPingTimedOutTimer()
+{
+    _pingTimeoutTimer.setInterval(_pingTimeoutInterval);
```

Comment:
@FlexW I'd name it **_pongTimeoutTimer** as all it does is waits for pong to arrive from the server within the specific interval. Also, I'd rename **_pingTimeoutInterval** -> **_pongTimeout** or **_pongWaitLimit**, as, in fact, the timer is not used for executing the specific call in a loop with an interval. It's just a single-shot timer that is set for the sake of waiting for the server's pong. If you know what I mean. The **interval** is something that makes me think of a network call that is performed infinitely. Like, for example, if the client is sending **heartbeat** to a server via UDP just to let it know that it's still alive. Now, the **ping** that we are sending - is indeed something that we are sending infinitely with an interval. In contrast to **wait for pong from server**.

Reply:
The pingTimeoutInterval is also used to schedule the ping. So every `pingTimeoutInterval` milliseconds the ping get executed. Naming it `pongWaitLimit` would be wrong in my opinion. We could just name it `pingInterval`. What do you think about this?